apps/cli/src/index.ts

```diff
@@ -121,10 +132,185 @@ program
+     * -- [DONE] filter out any .gitignore
+     * - [DONE] targz them up
+     * - [PARTIALLY DONE] make a POST request to METAL_URL + /api/deploy/up with:
+     * -- [DONE] the user's token as a bearer token in the Authorization header
+     * -- add the config for the current project too somehow so that we know the project we're deploying.
+     * - [DONE] upload the tarball. Once the upload is complete, return a tag to the command.
+     * - then this command calls GET METAL_URL + /api/deploy/{tag}/status to check the status of the deployment.
+     * - [DONE/IN-PROGRESS] if the deployment is ongoing, stream responses to the command line.
+     * - if it has already finished with a success or failure, return that and end this command.
+     */
+
+    let step = 1;
+    log(`[${step}] Checking for token...`);
+    const userConfig = checkUserConfig();
+    // Token hierachy: commandline > config file > environment variable
+    const token = options.token || userConfig.whoami.token || process.env.METAL_TOKEN;
+    if (!token) {
+      log("Error! You must configure a Metal API token.");
+      process.exit(1);
+    }
+
+    log(`[${++step}] Collating files to deploy...`);
+    const { stdout, stderr } = await exec(`git ls-files`);
+    if (stderr) {
+      console.error(stderr);
+      process.exit(1);
+    }
+
+    const pathsToArchive = stdout
+      .split("\n")
+      .filter((path) => !!path && !path.endsWith(".gitignore"));
+
+    log(`[${++step}] Compressing files...`);
+    const payloadStream = createTar({
+      gzip: true,
+      cwd: process.cwd(),
+    }, pathsToArchive);
+
+    log(`[${++step}] Uploading...`);
+    const reqOptions = {
+      host: "localhost",
+      port: 3000,
+      path: "/api/deploy/up",
+      method: "POST",
+      headers: {
+        "Content-Type": "application/octet-stream",
+        Authorization: `Bearer ${token}`,
+        "Accept": "application/json",
+      },
+    };
+
+    // TODO MET-10: explore any better ways of promisifying this.
+    const uploadPromise = new Promise<string>((resolve, reject) => {
+      const request = nodeRequest(reqOptions, response => {
+        let bodyJSONString = "";
+        response.on("data", (chunk) => {
+          bodyJSONString += chunk;
+        });
+
+        response.on("error", (err) => {
+          console.error(`Problem with response from upload: ${err.message}`);
+          reject(err);
+        });
+
+        response.on("end", () => {
+          resolve(bodyJSONString);
+        });
+      });
+
+      request.on("error", (err) => {
+        console.error(`Problem with upload request: ${err.message}`);
+        reject(err);
+      });
+
+      // Write compressed data to request's body
+      // TODO MET-10: consider using node:zlib
+      payloadStream.on("data", (chunk) => {
+        request.write(chunk);
+      });
+
+      payloadStream.on("end", () => {
+        request.end();
+      });
+    });
```

Comment:
is there any reason not to use bun's native `fetch` support here? I think the `Writable` and `Readable` interfaces can work with the tar old-style node stream. Here's what ChatGPT has to say:

```ts
const uploadData = async (payloadStream, token) => {
  const stream = new ReadableStream({
    start(controller) {
      payloadStream.on("data", chunk => {
        controller.enqueue(new Uint8Array(chunk)); // Assuming chunk is a Buffer this might be necessary
      });
      payloadStream.on("end", () => {
        controller.close();
      });
      payloadStream.on("error", err => {
        controller.error(err);
      });
    }
  });

  const reqOptions = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/octet-stream',
      'Authorization': `Bearer ${token}`,
      'Accept': 'application/json',
    },
    body: stream,
  };

  try {
    const response = await fetch('http://localhost:3000/api/deploy/up', reqOptions);
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    const bodyJSONString = await response.text();
    return JSON.parse(bodyJSONString);
  } catch (error) {
    console.error('Problem with upload request:', error.message);
    throw error;
  }
};
```

Reply:
- tar doesn't use old-style streams AFAIK. It uses some stream-like data type called `Pack`.
- Not sure this code is better. It essentially first pipes `payloadStream` into a readable stream and then gives that stream to the request, before then giving the new stream as the request's body. On the other hand, the existing code directly pipes the compressed data to the request's body.
- I think I chose Node's `request()` because of the ability to pipe to the stream object it returns. I'll try to switch to `fetch` instead and see how it goes.

Reply:
I gave it a shot in this commit https://github.com/onmetal-dev/metal/pull/16/commits/5235cd998995f01530e790bc39b27ebfbacd1cd8 but unfortunately couldn't get it to work. The uploaded file kept turning out to be zero bytes. I must be doing something wrong, but I don't know what to poke at. LMK when you see this if you notice what's missing.

Reply:
ok thanks for exploring this. I think this is fine, but some nits:

- I think it's confusing on first read to see a promise just defined on its own, so I think I prefer

```
    const bodyAsString = await new Promise<string>((resolve, reject) => {
```
to what it is currently


- `payloadStream.pipe(request);` accomplishes the same thing as the the "data" and "end" event handlers

Reply:
Done.